Arbitrary variants

RobinMalfait opened the pull request:
With the new `addVariant` API, we have a beautiful way of creating new variants.

You can use it as:
```js
addVariant('children', '& > *')
```

Now you can use the `children:` variant. The API uses a `&` as a reference for the candidate, which means that:
```html
children:pl-4
```

Will result in:
```css
.children\:pl-4 > * { .. }
```

Notice that the `&` was replaced by `.children\:pl-4`.

We can leverage this API to implement arbitrary variants, this means that you can write those `&>*` (Notice that we don't have spaces) inside a variant directly. An example of this can be:
```html
<ul class="[&>*]:underline">
  <li>A</li>
  <li>B</li>
  <li>C</li>
</ul>
```

Which generates the following css:
```css
.\[\&\>\*\]\:underline > * {
  text-decoration-line: underline;
}
```

Now all the children of the `ul` will have an `underline`. The selector itself is a bit crazy since it contains the candidate which is the selector itself, it is just escaped.

<!--

👋 Hey, thanks for your interest in contributing to Tailwind!

**Please ask first before starting work on any significant new features.**

It's never a fun experience to have your pull request declined after investing a lot of time and effort into a new feature. To avoid this from happening, we request that contributors create an issue to first discuss any significant new features. This includes things like adding new utilities, creating new at-rules, or adding new component examples to the documentation.

https://github.com/tailwindcss/tailwindcss/blob/master/.github/CONTRIBUTING.md

-->

RobinMalfait committed:
register arbitrary variants

With the new `addVariant` API, we have a beautiful way of creating new
variants.

You can use it as:
```js
addVariant('children', '& > *')
```

Now you can use the `children:` variant. The API uses a `&` as a
reference for the candidate, which means that:
```html
children:pl-4
```

Will result in:
```css
.children\:pl-4 > * { .. }
```

Notice that the `&` was replaced by `.children\:pl-4`.

We can leverage this API to implement arbitrary variants, this means
that you can write those `&>*` (Notice that we don't have spaces) inside
a variant directly. An example of this can be:
```html
<ul class="[&>*]:underline">
  <li>A</li>
  <li>B</li>
  <li>C</li>
</ul>
```
Which generates the following css:
```css
.\[\&\>\*\]\:underline > * {
  text-decoration-line: underline;
}
```

Now all the children of the `ul` will have an `underline`. The selector
itself is a bit crazy since it contains the candidate which is the
selector itself, it is just escaped.

RobinMalfait committed:
add tests for arbitrary variants

This still requires some work to the `defaultExtractor` to make sure it
all works with existing code.

RobinMalfait committed:
update changelog

RobinMalfait commented on src/lib/defaultExtractor.js:
```diff
@@ -19,7 +19,7 @@ export function defaultExtractor(content) {
   yield regex.pattern([
     // Variants
-    /((?=([^\s"'\\\[]+:))\2)?/,
+    /((?=([^\s"'\\]+:))\2)?/,
```
This will require some work to make it exactly right

thecrypticace replied:
lol yep I was thinking about this. I'll throw this change against my new candidate refactor to see what breaks…

I had already tested support for arbitrary variant syntax (at least detection wise) in the form of [foo]:bar

thecrypticace replied:
ngl I love this idea so much

thecrypticace replied:
I pushed a regex tweak to fix the tests 👍👍

thecrypticace committed:
Fix candidate detection for arbitrary variants

thecrypticace committed:
Refactor

thecrypticace committed:
Add support for at rules

simonswiss commented:
This is rad! 🤙

onmax commented:
Wow, just requested this feature about one month ago and now is coming here. Timing couldn't be better.  I have a few ideas in my head to create some plugins with this.

https://github.com/tailwindlabs/tailwindcss/discussions/8157

Keep up the hard work 🔥 🔥.  Amazing!

adamwathan committed:
Add test for attribute selectors

adamwathan commented:
Added a failing test for attribute selectors like:

`class="[&:[data-open]]:underline"`

I think this one is going to be kind of tricky — we need to reliable extract that candidate from the template but to do that we need to worry about bracket balancing. We'll need to support multiple pairs too so the implementation probably can't be as naive as I'd wish:

`class="[&:[data-foo][data-bar]:has([data-baz])]:underline"`

thecrypticace committed:
Fix test

thecrypticace committed:
Add attribute selector support

thecrypticace committed:
Split top-level comma parsing into a generalized splitting routine

We can now split on any character at the top level with any nesting. We don’t balance brackets directly here but this is probably “enough”

thecrypticace committed:
Split variants by separator at the top-level only

This means that the separator has to be ouside of balanced brackets

thecrypticace committed:
Fix extraction when using custom variant separators

thecrypticace committed:
Support custom separators when top-level splitting variants

thecrypticace committed:
Add a second multi-character separator test

adamwathan committed:
Split tests for at-rule and at-rule with selector changes

thecrypticace committed:
Add nested at-rule tests

thecrypticace committed:
Fix space-less at-rule parsing in addVariant

adamwathan committed:
Add test for using with `@apply`

jipika commented:
Is it possible to specify a class name? Sometimes it is not possible to change all of them

calvo-jp commented:
woah! how cool :rocket:

ivanchenhz commented:
Cool, 👍
And seems we don't support class selector?!

[&_i]:text-pink-100      👌
[&_i.some_class]:text-pink-100     ...

RobinMalfait commented:
@ivanchenhz we do, but `_` is used for spaces, so you essentially have this selector:
```css
i .some class {}
```
You have to escape the `_` if you want to keep it as an actual underscore. See: https://tailwindcss.com/docs/hover-focus-and-other-states#using-arbitrary-variants

ivanchenhz commented:
@RobinMalfait  thanks for confirming, and yes it works, 👍 😄
And I finally figure it out, the problem is we have prefix 'tw-'

- So, when I targeting class like ```$_i.tw-font-medium``` it generate ```i.tw-tw-font-medium```
- Should we considering DON'T count prefix in arbitrary mode !
<img width="581" alt="Screen Shot 2022-06-13 at 10 04 25 PM" src="https://user-images.githubusercontent.com/2640513/173371825-9f046a98-5821-43aa-9832-b85b030a75c4.png">


<img width="559" alt="Screen Shot 2022-06-13 at 10 04 12 PM" src="https://user-images.githubusercontent.com/2640513/173371738-3d513598-3c4f-46bf-9b11-a182e8e21d4d.png">

ivanchenhz commented:
@RobinMalfait 
Got a real case for this, we like to restyling the 'popover-body' which comes from some library
After add ```[&_.popover-body]:tw-pr-2```, it auto prefix to generate ```tw-popover-body``` ( doesn't actually exist )
Have to manually add the class to make it work, ( which not feasible obviously  😂 ）
Please considering don't count prefix in arbitrary mode


<img width="1169" alt="Screen Shot 2022-06-13 at 11 07 57 PM" src="https://user-images.githubusercontent.com/2640513/173385618-e59dcfb8-9650-401f-adba-dcc14a908878.png">

Emre-Zero commented:
Nice work!

I suppose can shorten to `ch`

In `tailwind.config.js`:
```javascript
const plugin = require("tailwindcss/plugin");
plugins: [
      plugin(function({ addVariant }) {
        addVariant('ch', '& > *')
      })
]
```

clintfisher commented:
Having issues with Arbitrary variants not showing in Intellisense:
https://github.com/tailwindlabs/tailwindcss/discussions/9449
